Added name field to both custom project and org roles

upodroid opened the pull request:
I updated the custom roles to export the name attribute so it can be used with iam role bindings.

I also updated the project_iam_policy article to include a note about importing the policy first before applying the change.

Fixes: https://github.com/terraform-providers/terraform-provider-google/issues/6089

https://cloud.google.com/iam/docs/reference/rest/v1/projects.roles#resource:-role

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
iam: Added `name` field to `google_project_iam_custom_role`
iam: Added `name` field to `google_organization_iam_custom_role`
```

upodroid commented:
@googlebot I signed it!

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.

Thanks for your contribution!  A human will be with you soon.

@c2thorn, please review this PR or find an appropriate assignee.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3370-old..auto-pr-3370) ( 5 files changed, 9 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3370-old..auto-pr-3370) ( 5 files changed, 9 insertions(+), 2 deletions(-))

c2thorn commented on third_party/terraform/resources/resource_google_organization_iam_custom_role.go:
```diff
@@ -122,6 +122,7 @@ func resourceGoogleOrganizationIamCustomRoleRead(d *schema.ResourceData, meta in
 	d.Set("role_id", parsedRoleName.Name)
 	d.Set("org_id", parsedRoleName.OrgId)
 	d.Set("title", role.Title)
+	d.Set("name", role.Name)
```
`name` won't appear in state until the field is established as part of the schema. It will need to be set to `Computed: true` as well.

upodroid replied:
Do I set this in the schema definition on resourceGoogleProjectIamCustomRole function?

c2thorn replied:
Yep, within the schema map. Just needs to be set to a string and computed.

c2thorn commented on third_party/terraform/website/docs/r/google_organization_iam_custom_role.html.markdown:
```diff
@@ -60,6 +60,8 @@ exported:
 
 * `deleted` - (Optional) The current deleted state of the role.
 
+* `name` - The name of the role which can be used with iam role bindings in the format `organizations/{{org_id}}/roles/{{role_id}}`
```
Let's separate out the definition and make the recommendation a bit more generic. Something like:
```suggestion
* `name` - The name of the role in the format `organizations/{{org_id}}/roles/{{role_id}}`. Like `id`, this field can be used as a reference in other resources such as IAM role bindings.
```

c2thorn commented on third_party/terraform/website/docs/r/google_project_iam.html.markdown:
```diff
@@ -28,7 +28,8 @@ Four different resources help you manage your IAM policy for a project. Each of
    It's not recommended to use `google_project_iam_policy` with your provider project
    to avoid locking yourself out, and it should generally only be used with projects
-   fully managed by Terraform.
+   fully managed by Terraform. If you do use this resource, **import** the policy before
```
```suggestion
   fully managed by Terraform. If you do use this resource, it is recommended to **import** the policy before
```

c2thorn replied:
nitpick here

c2thorn commented on third_party/terraform/website/docs/r/google_project_iam_custom_role.html.markdown:
```diff
@@ -49,7 +49,7 @@ The following arguments are supported:
 * `stage` - (Optional) The current launch stage of the role.
     Defaults to `GA`.
-    List of possible stages is [here](https://cloud.google.com/iam/reference/rest/v1/organizations.roles#Role.RoleLaunchStage).
+    List of possible stages is [here](https://cloud.google.com/iam/reference/rest/v1/projects.roles#Role.RoleLaunchStage).
```
Looks like only the "organizations" version of the link works, this version doesn't point to role launch stages

c2thorn commented on third_party/terraform/website/docs/r/google_project_iam_custom_role.html.markdown:
```diff
@@ -60,6 +60,8 @@ exported:
 
  * `deleted` - (Optional) The current deleted state of the role.
 
+ * `name` - The name of the role which can be used with iam role bindings in the format `projects/{{project}}/roles/{{role_id}}`
```
See the suggestion for `google_organization_iam_custom_role`

c2thorn commented on third_party/terraform/resources/resource_google_project_iam_custom_role.go:
```diff
@@ -126,6 +126,7 @@ func resourceGoogleProjectIamCustomRoleRead(d *schema.ResourceData, meta interfa
 
 	d.Set("role_id", GetResourceNameFromSelfLink(role.Name))
 	d.Set("title", role.Title)
+	d.Set("name", role.Name)
```
Same as the previous comment

upodroid commented:
I fixed all the issues in the comments. I also added the id field to make it clear. I knew the about the id field because I heard that it will be replacing selflinks in the future.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3370-old..auto-pr-3370) ( 5 files changed, 20 insertions(+), 1 deletion(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3370-old..auto-pr-3370) ( 5 files changed, 20 insertions(+), 1 deletion(-))

c2thorn commented on third_party/terraform/resources/resource_google_organization_iam_custom_role.go:
```diff
@@ -55,6 +55,10 @@ func resourceGoogleOrganizationIamCustomRole() *schema.Resource {
 				Computed: true,
 			},
+			"name": {
+				Type:     schema.TypeBool,
```
```suggestion
				Type:     schema.TypeString,
```

c2thorn commented on third_party/terraform/resources/resource_google_project_iam_custom_role.go:
```diff
@@ -58,6 +58,10 @@ func resourceGoogleProjectIamCustomRole() *schema.Resource {
 				Computed: true,
 			},
+			"name": {
+				Type:     schema.TypeBool,
```
```suggestion
				Type:     schema.TypeString,
```

upodroid commented:
Oopsie, copy and paste

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3370-old..auto-pr-3370) ( 5 files changed, 20 insertions(+), 1 deletion(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3370-old..auto-pr-3370) ( 5 files changed, 20 insertions(+), 1 deletion(-))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3370-old..auto-pr-3370) ( 5 files changed, 20 insertions(+), 1 deletion(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3370-old..auto-pr-3370) ( 5 files changed, 20 insertions(+), 1 deletion(-))
